cmd/daemon.go

```diff
@@ -0,0 +1,119 @@
+package cmd
+
+import (
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+var (
+	done                            chan bool
+	controlChannels, returnChannels []chan bool
+)
+
+// daemonCmd updates Certs from vault to the end of time
+var daemonCmd = &cobra.Command{
+	Use:   "daemon",
+	Short: "Run forever",
+	Long:  `Work as daemon updating the certificates as needed`,
+	Run: func(cmd *cobra.Command, args []string) {
+		signals := make(chan os.Signal, 1)
+		signal.Notify(signals, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM)
+		go handleSignals(signals)
+		startWatchingCertificates()
+		<-done
+		for _, returnChannel := range returnChannels {
+			<-returnChannel
+		}
+		log.Info("exiting")
+	},
+}
+
+func init() {
+	updateCmd.AddCommand(daemonCmd)
+	done = make(chan bool, 1)
+}
+
+func handleSignals(signalChannel chan os.Signal) {
+	for {
+		signal := <-signalChannel
+		log.Infof("caught signal: %s", signal)
+		switch signal {
+		case syscall.SIGHUP:
+			reloadConfig()
+			startWatchingCertificates()
+		case syscall.SIGINT, syscall.SIGTERM:
+			done <- true
+			return
+		default:
+		}
+	}
+}
+
+func reloadConfig() {
+	// Stop existing functions
+	for _, controlChannel := range controlChannels {
+		controlChannel <- true
+	}
+	// make sure we're done
+	for _, returnChannel := range returnChannels {
+		<-returnChannel
+	}
+	// Clean up configurations
+	controlChannels = nil
+	returnChannels = nil
+	Conf = nil
+	// Reload configuration
+	initConfig()
+}
+
+func startWatchingCertificates() {
+	readInCertificates()
+	for _, cconf := range Conf.Certificates {
+		controlChannel := make(chan bool, 1)
+		returnChannel := make(chan bool, 1)
+		log.Infof("starting watching for: %s", cconf.CName)
+		go cconf.updateInvalidCertificateAdInfinitum(controlChannel, returnChannel)
+		controlChannels = append(controlChannels, controlChannel)
+		returnChannels = append(returnChannels, returnChannel)
+	}
+}
+
+func (cconf *CertificateConf) updateInvalidCertificateAdInfinitum(controlChannel, returnChannel chan bool) {
+	for {
+		cconf.updateInvalidCertificate()
+		defer func() {
+			log.Infof("writing files for certificate: %s", cconf.CName)
+			if err := cconf.WriteFiles(); err != nil {
```

Comment:
So, I'm probably missing something, but I see that ``UpdateInvalidCertificate()`` in update.go already calls ``WriteFiles()`` (currently line 59), so why are we doing it again?

Reply:
Also why is the call ``cconf.updateInvalidCertificate()`` outside of the defer but ``cconf.WriteFiles()`` is inside?

Reply:
I was getting paranoid about the concurrency. But I think you're right. I'll ditch this.